src/renderer/renderElement.ts

```diff
@@ -300,6 +311,30 @@ const drawElementOnCanvas = (
   context.globalAlpha = 1;
 };
 
+type LineGroupedRanges = { color: string; text: string }[][];
+
+const getLineGroupedRanges = (
+  element: ExcalidrawTextElement,
+): LineGroupedRanges => {
+  const lines: LineGroupedRanges = [];
+
+  let charCount = 0;
+  element.text
+    .replace(/\r\n?/g, "\n")
+    .split("\n")
+    .forEach((elementLine) => {
+      lines.push(
+        [...elementLine].map((char, i) => ({
```

Comment:
We have a problem with grapheme clusters or multi-codepoint-letters or whatever you want to call them. The core issue with the current code is that  `[...'🤌🏼👌🏿']` evaluates to `[ '🤌', '🏼', '👌', '🏿' ]` 

https://user-images.githubusercontent.com/9257001/167246490-05ecd607-74e9-4d6d-ab9a-741cbb0d01a6.mov

Reply:
I'm assuming that the text selection API also works on graphemes ie. selecting `🤌🏼` gives you a range of length 1 not 2. If this is the case I think the only solution is to have both `textElement.colorRanges` and `getLineGroupedRanges` work off of graphemes too.

In any case somewhere is going to have to be responsible for splitting by grapheme which probably means we need to pull in https://github.com/orling/grapheme-splitter

Reply:
Ah, we should use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter

Reply:
haven't reviewed yet so unclear what issues you're facing exactly, but do we even need to be aware of graphemes? Can't we just save the exact offsets of a given selection range? I assume the answer is we can't mainly due to how it would work alongside line breaks when bound text comes into play?

If not, then the key is not to use array spread `[...string]` because it splits on code points (thus is potentially multi-byte), and instead use `string.split('')` which splits on bytes.

Reply:
The problem with doing `.split('')` here is that once they're split it's not clear how to recombine the codepoints to form the graphemes. This is also why we need to be aware of graphemes in the first place, not sure you can apply a colour range without worrying about colour ranges because a range applies to graphemes - not codepoints.

Reply:
Depending on what you do afterwards, you should still be able to treat the string as a black box and it won't really matter?

For example, if you have `Peru 🇵🇪`, and the user selected the flag emoji, the range would be `[5, 9]` (the flag is 4 bytes). You split which gets you:

```
const bytes = ['P', 'e', 'r', 'u', ' ', '\uD83C', '\uDDF5', '\uD83C', '\uDDEA']
```

You extract the flag by slicing it out:

```
const selection = bytes.slice(5, 9);
```

You can then `selection.join('')` which gets you the 🇵🇪 grapheme back.

Reply:
Hmmm, for some reason I had the idea that `str.split('').join('') === str` wasn't always true.

Reply:
Seems to work fine in practice though and this suggestion definitely makes sense intuitively. I've implemented it in e7d5bfa37d17ff38a5a29793a7188eca153c0f90